AndroidSDKTests/src/test/java/com/leanplum/internal/RequestTest.java

```diff
@@ -198,12 +202,112 @@ public void testRemoveIrrelevantBackgroundStartRequests() throws NoSuchMethodExc
       put(Constants.Params.BACKGROUND, Boolean.toString(true));
       put("bg", "2");
     }}).sendEventually();
-    unsentRequests = (List) getUnsentRequests.invoke(Request.class);
+    unsentRequests = request.getUnsentRequests(1.0);
     unsentRequestsData =
         (List) removeIrrelevantBackgroundStartRequests.invoke(Request.class, unsentRequests);
     assertNotNull(unsentRequestsData);
     assertEquals(3, unsentRequestsData.size());
     Request.deleteSentRequests(unsentRequests.size());
     LeanplumEventDataManagerTest.setDatabaseToNull();
   }
+
+  @Test
+  public void testJsonEncodeUnsentRequestsWithExceptionLargeNumbers() throws NoSuchMethodException,
+          InvocationTargetException, IllegalAccessException {
+    LeanplumEventDataManager.init(Leanplum.getContext());
+    Request.RequestsWithEncoding requestsWithEncoding;
+    // Prepare testable objects and method.
+    Request request = spy(new Request("POST", Constants.Methods.START, null));
+    request.sendEventually();
+    for (int i = 0;i < 4999; i++) {
+      new Request("POST", Constants.Methods.START, null).sendEventually();
+    }
+    // Expectation: 5000 requests returned.
+    requestsWithEncoding = request.getRequestsWithEncodedStringStoredRequests(1.0);
+
+    assertNotNull(requestsWithEncoding.unsentRequests);
+    assertNotNull(requestsWithEncoding.requestsToSend);
+    assertNotNull(requestsWithEncoding.jsonEncodedString);
+    assertEquals(5000, requestsWithEncoding.unsentRequests.size());
+
+    // Throw OOM on 5000 requests
+    // Expectation: 2500 requests returned.
+    when(request.getUnsentRequests(1.0)).thenThrow(OutOfMemoryError.class);
+    requestsWithEncoding = request.getRequestsWithEncodedStringStoredRequests(1.0);
+
+    assertNotNull(requestsWithEncoding.unsentRequests);
+    assertNotNull(requestsWithEncoding.requestsToSend);
+    assertNotNull(requestsWithEncoding.jsonEncodedString);
+    assertEquals(2500, requestsWithEncoding.unsentRequests.size());
+
+    // Throw OOM on 2500, 5000 requests
+    // Expectation: 1250 requests returned.
+    when(request.getUnsentRequests(0.5)).thenThrow(OutOfMemoryError.class);
+    requestsWithEncoding = request.getRequestsWithEncodedStringStoredRequests(1.0);
+    assertEquals(1250, requestsWithEncoding.unsentRequests.size());
+
+    // Throw OOM on > 0 requests
+    // Expectation: 0 requests returned but no crash
```

Comment:
If I am understanding right 50% of requests are returned during OOM but don't crash